index.html

```diff
@@ -204,55 +204,1951 @@ <h3>
+        </p>
+        <ol>
+          <li>Let |promise:Promise| be [=a new promise=].
+          </li>
+          <li>If [=this=].{{HIDDevice/[[state]]}} is not `"closed"`, [=reject=]
+          |promise| with a "{{InvalidStateError}}" {{DOMException}} and abort
+          these steps.
+          </li>
+          <li>Set [=this=].{{HIDDevice/[[state]]}} to `"opening"`.
+          </li>
+          <li>Perform the following steps [=in parallel=]:
+            <ol>
+              <li>Invoke the operating system to open the HID device.
+              </li>
+              <li>If this fails for any reason, [=queue a global task=] on the
+              [=relevant global object=] of [=this=] using the [=HID device
+              task source=] to [=reject=] |promise| with a
+              "{{NotAllowedError}}" {{DOMException}} and abort these steps.
+              </li>
+              <li>Set [=this=].{{HIDDevice/[[state]]}} to `"opened"`.
+              </li>
+              <li>[=Queue a global task=] on the [=relevant global object=] of
+              [=this=] using the [=HID device task source=] to [=resolve=]
+              |promise| with `undefined`.
+              </li>
+            </ol>
+          </li>
+          <li>Return |promise|.
+          </li>
+        </ol>
+      </section>
+      <section>
+        <h3>
+          <dfn>close()</dfn> method
+        </h3>
+        <p>
+          The {{HIDDevice/close()}} method steps are:
+        </p>
+        <ol>
+          <li>Let |promise:Promise| be [=a new promise=].
+          </li>
+          <li>Set [=this=].{{HIDDevice/[[state]]}} to `"closing"`.
+          </li>
+          <li>Perform the following steps [=in parallel=]:
+            <ol>
+              <li>Invoke the operating system to close the HID device and
+              release any associated resources.
+              </li>
+              <li>Set [=this=].{{HIDDevice/[[state]]}} to `"closed"`.
+              </li>
+              <li>[=Queue a global task=] on the [=relevant global object=] of
+              [=this=] using the [=HID device task source=] to [=resolve=]
+              |promise| with `undefined`.
+              </li>
+            </ol>
+          </li>
+          <li>Return |promise|.
+          </li>
+        </ol>
+      </section>
+      <section>
+        <h3>
+          <dfn>sendReport()</dfn> method
+        </h3>
+        <p>
+          The {{HIDDevice/sendReport()}} method steps are:
+        </p>
+        <ol>
+          <li>Let |promise:Promise| be [=a new promise=].
+          </li>
+          <li>If [=this=].{{HIDDevice/[[state]]}} is not `"opened"`, [=reject=]
+          |promise| with a "{{InvalidStateError}}" {{DOMException}} and abort
+          these steps.
+          </li>
+          <li>Perform the following steps [=in parallel=]:
+            <ol>
+              <li>Invoke the operating system to send an output report with the
+              specified |reportId:octet| and |data:BufferSource|.
```

Comment:
Add a note on passing 0 if the device does not use report IDs. Should that throw an error?

Reply:
The [DCD for HID](https://www.usb.org/document-library/device-class-definition-hid-111) says "Report ID zero is reserved and should not be used."

I'm reluctant to validate report IDs since it's not uncommon for devices to violate the HID spec in ways that don't prevent access through the platform HID APIs. I think in general we shouldn't reject a report at the WebHID level if it could have been delivered successfully by the platform API. Once the platform API call fails we'll reject with NotAllowedError.

With that in mind, it's probably fine to reject non-zero report IDs for devices that don't use report IDs and report ID 0 for devices that do use them since these reports would be rejected by the platform HID APIs anyway. Let's check for these cases and reject with TypeError.